wire/msgmixsecrets.go

```diff
@@ -0,0 +1,248 @@
+	err = readElement(r, &numSRs)
+	if err != nil {
+		return err
+	}
+	if numSRs > MaxMixMcount {
+		msg := fmt.Sprintf("too many total mixed messages [count %v, max %v]",
+			numSRs, MaxMixMcount)
+		return messageError(op, ErrInvalidMsg, msg)
+	}
+	msg.SlotReserveMsgs = make([][]byte, numSRs)
+	for i := uint64(0); i < numSRs; i++ {
+		sr, err := ReadVarBytes(r, pver, MaxMixFieldValLen, "SR")
+		if err != nil {
+			return err
+		}
+		msg.SlotReserveMsgs[i] = sr
+	}
+
+	var numMs uint64
+	err = readElement(r, &numMs)
+	if err != nil {
+		return err
+	}
+	if numMs > MaxMixMcount {
+		msg := fmt.Sprintf("too many total mixed messages [count %v, max %v]",
+			numMs, MaxMixMcount)
+		return messageError(op, ErrInvalidMsg, msg)
+	}
+	msg.DCNetMsgs = make([][]byte, numMs)
+	for i := uint64(0); i < numMs; i++ {
+		m, err := ReadVarBytes(r, pver, MaxMixFieldValLen, "M")
+		if err != nil {
+			return err
+		}
+		msg.DCNetMsgs[i] = m
+	}
+
+	return nil
+}
+
+// BtcEncode encodes the receiver to w using the Decred protocol encoding.
+// This is part of the Message interface implementation.
+func (msg *MsgMixSecrets) BtcEncode(w io.Writer, pver uint32) error {
+	const op = "MsgMixSecrets.BtcEncode"
+	if pver < MixVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	err := writeElement(w, &msg.Signature)
+	if err != nil {
+		return err
+	}
+
+	err = msg.writeMessageNoSignature(op, w, pver)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Hash returns the message hash calculated by WriteHash.
+//
+// Hash returns an invalid or zero hash if WriteHash has not been called yet.
+//
+// This method is not safe while concurrently calling WriteHash.
+func (msg *MsgMixSecrets) Hash() chainhash.Hash {
+	return msg.hash
+}
+
+// WriteHash serializes the message to a hasher and records the sum in the
+// message's Hash field.
+//
+// The hasher's Size() must equal chainhash.HashSize, or this method will
+// panic.  This method is designed to work only with hashers returned by
+// blake256.New.
+func (msg *MsgMixSecrets) WriteHash(h hash.Hash) {
+	h.Reset()
+	writeElement(h, &msg.Signature)
+	msg.writeMessageNoSignature("", h, MixVersion)
+	sum := h.Sum(msg.hash[:0])
+	if len(sum) != len(msg.hash) {
+		s := fmt.Sprintf("hasher type %T has invalid Size() for chainhash.Hash", h)
+		panic(s)
+	}
+}
+
+// writeMessageNoSignature serializes all elements of the message except for
+// the signature.  This allows code reuse between message serialization, and
+// signing and verifying these message contents.
+//
+// This method never errors for invalid message construction.
+func (msg *MsgMixSecrets) writeMessageNoSignature(op string, w io.Writer, pver uint32) error {
+	err := writeElements(w, &msg.Identity, &msg.SessionID, msg.Expiry,
+		msg.Run, &msg.Seed)
+	if err != nil {
+		return err
+	}
+
+	err = writeElement(w, uint64(len(msg.SlotReserveMsgs)))
```

Comment:
This should be a varint imo.  Given the max count here is 1024, that's only 3 bytes for the varint versus 8 for the `uint64`.  Naturally, the read side and max payload length would need to change accordingly too.

Reply:
Total slice length for this field is written as a varint now.